### PR TITLE
Feature/refactor

### PR DESCRIPTION
**Pull request by david-macleod**

### CODE REVIEW

File: Fast and Accurate GPU Quantization for Transformers.md [hunk 1]
```diff
@@ -1,25 +1,10 @@
...
 
-This has led to a surge in interest for optimizing transformer inference. There are a range of techniques available [1], including:
+One popular technique for this is **quantization** - by reducing the precision of network parameters and activations, quantization aims to **increase throughput** and **decrease memory footprint**. Whilst reducing peak memory by quantizing from FP16 to INT8 is pretty much guaranteed, achieving performance improvements can be trickier. This is particularly true when running INT8 operations on GPU.
 
-* Sparsity
-* Pruning
-* Conditional Computation (e.g., Mixture of Experts)
-* Knowledge Distillation
-* Quantization
+The main challenge stems from the overheads associated with the quantization/dequantization operations - which can mask the performance gains from the blazingly fast INT8 Tensor Core matrix multiplies. The second part of this blog explores the nuances around achieving peak performance on GPU, and can be found here.
```
Comment (Lawrencium77) [inline Fast and Accurate GPU Quantization for Transformers.md:5]: It seems strange to me that we introduce the second half of the blog before the first half. Is there a way to swap them around?

Comment (Lawrencium77) [reply Fast and Accurate GPU Quantization for Transformers.md:5]: Also, I'm guessing you want to add a link at the end of this sentence?

File: Fast and Accurate GPU Quantization for Transformers.md [hunk 2]
```diff
@@ -1,25 +1,10 @@
...
-
-Provided the decrease in accuracy is minimal, this sounds ideal. Indeed, it is relatively easy to quantize a model in a way that reduces memory footprint and preserves accuracy. However, there remains an issue:
-
-> Implementing a GPU-based quantization scheme that **actually accelerates inference** is not straightforward. 
-
-The main challenge stems from the overheads associated with converting between floating point (FP) and Integer (INT) data types. Furthermore, there is a lack of fast General Matrix Multiply (GEMM) GPU kernels for precisions lower than INT8. For this reason, we focus solely on INT8 quantization in this blog.
-
-We describe an approach to GPU-based quantization of transformers that is both **flexible and fast**. The content is organized as follows:
+Achieving high throughput and low memory is all very well, but largely useless if quantizing to lower precision results in degradations in model accuarcy. In the first part of this blog we will introduce the quantization formulation, and explore some techniques for preserving accuracy - *whilst maintaining high throughput*.
```
Comment (Lawrencium77) [inline Fast and Accurate GPU Quantization for Transformers.md]:
```suggestion
Achieving high throughput and low memory is all very well, but largely useless if quantizing to lower precision results in degradations in model accuarcy. In the first part of this blog we will introduce the quantization formulation, and explore some techniques for preserving accuracy - *while maintaining high throughput*.
```

File: Fast and Accurate GPU Quantization for Transformers.md [hunk 3]
```diff
@@ -38,53 +23,52 @@ We describe an approach to GPU-based quantization of transformers that is both *
 	* [Fusion Implementation](#Fusion%20Implementation)
 * [References](#References)
 
+# Part I: Accurate Quantization
 
-## Background
-
-We’ll begin with a super quick summary of quantization. For further reading on this subject, we’ve listed some nice blogs/papers in our [References](#References) section [1-4].
+We’ll begin with a quick overview of quantization. For deeper reading on this subject, some nice blogs/papers in the [References](#References) section [1-4].
```
Comment (Lawrencium77) [inline Fast and Accurate GPU Quantization for Transformers.md:28]: Not sure this reads well now?

File: Fast and Accurate GPU Quantization for Transformers.md [hunk 4]
```diff
@@ -38,53 +23,52 @@ We describe an approach to GPU-based quantization of transformers that is both *
...
+In principle, we can use any function to convert from a higher-precision to lower-precision representation. But a linear function is simplest and quickest [4]:
 
-Here, $Q, r$ are the INT output and FP input, while $S, Z$ are a scale factor and bias. $\textrm{Int}$ is a function that rounds to the nearest integer. To minimize inference cost, we can set $Z=0$ (we do this for our GPU implementation). 
+$$Q(x)=\textrm{Int}(x/S)-Z$$ 
 
-The corresponding dequantization equation is even simpler:
+Here, $Q$ and $x$ are the fixed-point output and floating-point input, while $S$ and $Z$ represent the scale factor and bias. $\textrm{Int}$ is a function that rounds to the nearest integer, clipping values outside of the representable range. 
 
-$$\tilde{r}=S Q(r)$$
+After applying our lower-precision operation we return the data to it's original dynamic range by dequantizing the data:
```
Comment (Lawrencium77) [inline Fast and Accurate GPU Quantization for Transformers.md]:
```suggestion
After applying our lower-precision operation we return the data to it's original dynamic range with dequantization: 
```

File: Fast and Accurate GPU Quantization for Transformers.md [hunk 5]
```diff
@@ -38,53 +23,52 @@ We describe an approach to GPU-based quantization of transformers that is both *
...
+After applying our lower-precision operation we return the data to it's original dynamic range by dequantizing the data: 
+
+$$\tilde{x}=S Q(x) + Z$$
 
 This method is called **uniform quantization** since the quantized values are uniformly distributed over the input space. To calculate $S$ we select a **clipping range** $[\alpha, \beta]$ and then use:
 
 $$S=\frac{\beta-\alpha}{2^b-1}$$
 
-Here, $b$ is the number of bits in our quantization scheme. We choose to enforce $\alpha=-\beta$, which is known as **symmetric quantization**. This simplifies the (de)quantization function by setting $Z=0$.
+Here, $b$ is the number of bits in our quantization scheme. GPU based quantization schemes typicall enforce $\alpha=-\beta$, which is known as **symmetric quantization**. This simplifies the (de)quantization functions by setting $Z=0$, which helps reduce the cost of the transformation {do we have any reference for this??}.
```
Comment (Lawrencium77) [inline Fast and Accurate GPU Quantization for Transformers.md]:
```suggestion
Here, $b$ is the number of bits in our quantization scheme. GPU based quantization schemes typically enforce $\alpha=-\beta$, which is known as **symmetric quantization**. This simplifies the (de)quantization functions by setting $Z=0$, which helps reduce the cost of the transformation [4](https://arxiv.org/pdf/2103.13630.pdf).
```

Comment (Lawrencium77) [reply Fast and Accurate GPU Quantization for Transformers.md]: Yes, we can use one of the references we already have: https://arxiv.org/pdf/2103.13630.pdf

File: Fast and Accurate GPU Quantization for Transformers.md [hunk 6]
```diff
@@ -96,16 +80,16 @@ It’s also feasible to share quantization parameters between some subgroups of
 
 ## Specifics of  INT8 GEMMs
 
-The core element of a quantized neural network is INT8 matrix multiplication. Focusing on its details is crucial for an efficient implementation. This section describes these details, and serves as context for the later section describing [Implementation](#Implementation).
+The core element of a quantized neural network is INT8 matrix multiplication. Understanding its details is crucial for an efficient implementation. This section describes these details, and serves as context for [Part II](Part II: Fast GPU Quantization in Practice) of the blog
```
Comment (Lawrencium77) [inline Fast and Accurate GPU Quantization for Transformers.md]:
```suggestion
The core element of a quantized neural network is INT8 matrix multiplication. Understanding its details is crucial for an efficient implementation. This section describes these details, and serves as context for [Part II](Part II: Fast GPU Quantization in Practice) of the blog.
```

File: Fast and Accurate GPU Quantization for Transformers.md [hunk 7]
```diff
@@ -185,27 +172,33 @@ A consequence of this is that SmoothQuant can only be applied to matrix multipli
...
 
-As such, for the remainder of the blog we will focus on the challenges of achieving both good INT8 performance _and_ flexibility, by creating modular components that can be applied to different architectures, whilst remaining within the PyTorch framework for the non-quantized components.
+Ideally we can achieve the peformance of the inference frameworks, but retaining the flexibility of the afforementioned libraries. For the remainder of the blog we will focus on the challenges of achieving both good INT8 performance _and_ flexibility, by proposing modular components that can be applied to different architectures, whilst remaining within the PyTorch framework for non-quantized components.
```
Comment (Lawrencium77) [inline Fast and Accurate GPU Quantization for Transformers.md]:
```suggestion
Ideally we can achieve the performance of the inference frameworks, but retaining the flexibility of the aforementioned libraries. For the remainder of the blog we will focus on the challenges of achieving both good INT8 performance _and_ flexibility, by proposing modular components that can be applied to different architectures, whilst remaining within the PyTorch framework for non-quantized components.
```

File: Fast and Accurate GPU Quantization for Transformers.md (hunk 1 above)
Comment (david-macleod) [reply Fast and Accurate GPU Quantization for Transformers.md:5]: Yeah this is a bit weird, it came from an earlier version where I had split it into two separate blogs and made more sense there. I basically wanted to make it obvious that they could be read almost independently for people that understand quantization already. Will try and rework it

File: Fast and Accurate GPU Quantization for Transformers.md (hunk 3 above)
Comment (david-macleod) [reply Fast and Accurate GPU Quantization for Transformers.md:28]: Agree, lets revisit after merging

File: Fast and Accurate GPU Quantization for Transformers.md (hunk 1 above)
Comment (david-macleod) [reply Fast and Accurate GPU Quantization for Transformers.md:5]: Lets review this on main branch

**Pull request merged**